packages/back-end/src/services/openai.ts

```diff
@@ -0,0 +1,143 @@
+import { Configuration, OpenAIApi, ChatCompletionRequestMessage } from "openai";
+import { encoding_for_model, get_encoding } from "@dqbd/tiktoken";
+import { logger } from "../util/logger";
+import { OrganizationInterface } from "../../types/organization";
+import {
+  getTokenUsedByOrganization,
+  updateTokenUsage,
+} from "../models/AITokenUsageModel";
+
+/**
+ * Snapshot of gpt-3.5-turbo from March 1st 2023. Unlike gpt-3.5-turbo, this
+ * model will not receive updates, and will be deprecated 3 months after a new
+ * version is released.
+ *
+ * We use this model to ensure behavior doesn't change while gpt-3.5-turbo is
+ * updated. Additionally, token counts will be more predictable.
+ */
+const MODEL = "gpt-3.5-turbo-0301";
+
+/**
+ * The MODEL_TOKEN_LIMIT is the maximum number of tokens that can be sent to
+ * the OpenAI API in a single request. This limit is imposed by OpenAI.
+ *
+ * Note too that very long conversations are more likely to receive incomplete
+ * replies. For example, a gpt-3.5-turbo conversation that is 4090 tokens long
+ * will have its reply cut off after just 6 tokens.
+ */
+const MODEL_TOKEN_LIMIT = 4096;
+// Require a minimum of 30 tokens for responses.
+const MESSAGE_TOKEN_LIMIT = MODEL_TOKEN_LIMIT - 30;
+// TODO fine tune
+const DAILY_TOKEN_LIMIT = 1000000;
+
+const configuration = new Configuration({
+  apiKey: process.env.OPENAI_API_KEY || "",
+});
+
+const openai = new OpenAIApi(configuration);
```

Comment:
Can we make this a singleton instead and lazy load it the first time it's used?  This code now will run for everyone, even those without API keys.  It might not throw an error when apiKey is empty today, but I could totally see them adding that validation later

Reply:
Yes, good catch